### PR TITLE
FIX(Server): Non-Client TCP connections no longer add to user count

### PR DESCRIPTION
**Pull request by LucasToole**
This commit fixes the issues brought up in #4277 which were both
related to the fact that any TCP connection to a Mumble server would
be counted as a user. This allowed for an easy Denial of Service
attack even if the server is password protected, by only needing a
simple script making $MAX_USERS connections.

To fix this 2 problems needed to be addressed:
1) Only authenticated clients should add to the user count
2) The number of potential client IDs should not be dependent on the
max amount of users.

For the first issue the variable iAuthenticatedUsers keeps track of
the number of Authenticated Users currently connected. It is only
updated when the Voice Thread is write-locked. The server uses this
value to determine if a user can connect and advertises this value to
pings.

For the second issue, the concept of a iMaxUser dependent, size-capped ID Queue is
scraped in favor of unique, randomly generated IDs. To achieve this
the in-repo CryptographicRandom::uint32_t() function is used to avoid
external dependencies. When generating an ID, its first checked that
the ID isn't already in use before applying. The likelihood of ID
collisions with 2^32 possibilities is incredibly low, and it's
assumed that if your server is sustaining enough connections for
constant ID collisions you have further issues to address.

FIXES: #4277

### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

### CODE REVIEW

File: src/murmur/Messages.cpp [hunk 1]
```diff
@@ -1862,8 +1863,8 @@ void Server::msgContextAction(ServerUser *uSource, MumbleProto::ContextAction &m
 /// @param str The std::string to convert
 /// @param maxSize The maximum allowed size for this string
 /// @returns The given std::string converted to a QString, if its size is less
-///  	than or equal to the given maxSize. If it is bigger, "[[Invalid]]"
-///  	is returned.
+///	than or equal to the given maxSize. If it is bigger, "[[Invalid]]"
+///	is returned.
```
Comment (Krzmbrzl) [inline src/murmur/Messages.cpp:1867]: Unrelated change

File: src/murmur/Server.cpp [hunk 2]
```diff
@@ -721,7 +716,7 @@ void Server::udpActivated(int socket) {
 	socklen_t fromlen = sizeof(from);
 	int &sock         = socket;
 	len               = static_cast< qint32 >(
-        ::recvfrom(sock, encrypt, UDP_PACKET_SIZE, MSG_TRUNC, reinterpret_cast< struct sockaddr * >(&from), &fromlen));
+				  ::recvfrom(sock, encrypt, UDP_PACKET_SIZE, MSG_TRUNC, reinterpret_cast< struct sockaddr * >(&from), &fromlen));
```
Comment (Krzmbrzl) [inline src/murmur/Server.cpp:719]: Unrelated change

File: src/murmur/Server.cpp [hunk 3]
```diff
@@ -743,7 +738,7 @@ void Server::udpActivated(int socket) {
 		iov[0].iov_len = 6 * sizeof(quint32);
 		::sendmsg(sock, &msg, 0);
 #else
-        ::sendto(sock, encrypt, 6 * sizeof(quint32), 0, reinterpret_cast< struct sockaddr * >(&from), fromlen);
+		::sendto(sock, encrypt, 6 * sizeof(quint32), 0, reinterpret_cast< struct sockaddr * >(&from), fromlen);
```
Comment (Krzmbrzl) [inline src/murmur/Server.cpp:741]: Unrelated change, but seems to make sense as apparently the indentation was done using spaces instead of tabs before

**Review by Krzmbrzl — COMMENTED** [review]
The change of only counting authenticated users and the one of using a random user session (ID) seem unrelated to me. If that's the case, then these should be separate commits (no need for a new PR though).

What is the reasoning for changing the ID distribution in the first place? I think I have never encountered a problem with it (which doesn't have to mean that there is none) :thinking: 

EDIT: The second commit containing the new ID approach should probably be labeled `FEAT(server)` unless there is an actual bug that this addresses

**Comment by LucasToole** [comment]
I initially decided to keep both changes as one commit because they were both needed to fix the issue. 

The second change is one solution to the two issues that cause ID exhaustion.

1. The max amount of IDs a server has is (iMaxUsers * 2) - 1.
2. Murmur assumes that all connections are potential clients and always hands out IDs.

Therefore without addressing at least one of these issues, the connections needed to lock users out of a server goes from iMaxUsers -> (iMaxUsers * 2) - 1.

I had originally gone with solution 1, and changed the IDs from a very limited pool of in-order integers (until someone disconnects out of order) to randomly generated unique IDs.

However after sleeping on it I revisited option 2 which I had initially thought as harder and found that Murmur doesn't seem to require that connections need IDs until they are authenticated. So I'm adding a simpler patch which just move ID assignment to after a connection is authenticated and it looks like it solve both problems.

I still believe that assigning IDs in a stateless manner is preferable to the current ID Queue Approach. Maybe it will make its way into a future patch.
